src/algebra/lie/weights.lean

```diff
@@ -0,0 +1,253 @@
+
+  /- Unpack the information we have about `m₁`, `m₂`. -/
+  simp only [mem_pre_weight_space] at hm₁ hm₂,
+  obtain ⟨k₁, hk₁⟩ := hm₁ x,
+  obtain ⟨k₂, hk₂⟩ := hm₂ x,
+  have hf₁ : (f₁^k₁) (m₁ ⊗ₜ m₂) = 0,
+  { simp only [hk₁, zero_tmul, linear_map.rtensor_tmul, linear_map.rtensor_pow], },
+  have hf₂ : (f₂^k₂) (m₁ ⊗ₜ m₂) = 0,
+  { simp only [hk₂, tmul_zero, linear_map.ltensor_tmul, linear_map.ltensor_pow], },
+
+  /- It's now just an application of the binomial theorem. -/
+  use k₁ + k₂ - 1,
+  have hf_comm : commute f₁ f₂,
+  { ext m₁ m₂,
+    simp only [smul_comm (χ₁ x) (χ₂ x), linear_map.rtensor_smul, to_endomorphism_apply_apply,
+      tensor_product.mk_apply, linear_map.mul_apply, linear_map.one_apply, linear_map.ltensor_sub,
+      linear_map.compr₂_apply, linear_map.smul_apply, linear_map.ltensor_smul,
+      linear_map.rtensor_tmul, linear_map.map_sub, linear_map.map_smul, linear_map.rtensor_sub,
+      linear_map.sub_apply, linear_map.ltensor_tmul], },
+  rw hf_comm.add_pow',
+  simp only [tensor_product.map_incl, submodule.subtype_apply, finset.sum_apply,
+    submodule.coe_mk, linear_map.coe_fn_sum, tensor_product.map_tmul, linear_map.smul_apply],
+
+  /- The required sum is zero because each individual term is zero. -/
+  apply finset.sum_eq_zero,
+  rintros ⟨i, j⟩ hij,
+
+  /- Eliminate the binomial coefficients from the picture. -/
+  suffices : (f₁^i * f₂^j) (m₁ ⊗ₜ m₂) = 0, { rw this, apply smul_zero, },
+
+  /- Finish off with appropriate case analysis. -/
+  cases nat.le_or_le_of_add_eq_add_pred (finset.nat.mem_antidiagonal.mp hij) with hi hj,
+  { rw [(hf_comm.pow_pow i j).eq, linear_map.mul_apply, linear_map.pow_map_zero_of_le hi hf₁,
+    linear_map.map_zero], },
+  { rw [linear_map.mul_apply, linear_map.pow_map_zero_of_le hj hf₂, linear_map.map_zero], },
+end
+
+variables {M}
+
+lemma lie_mem_pre_weight_space_of_mem_pre_weight_space {χ₁ χ₂ : L → R} {x : L} {m : M}
+  (hx : x ∈ pre_weight_space L χ₁) (hm : m ∈ pre_weight_space M χ₂) :
+  ⁅x, m⁆ ∈ pre_weight_space M (χ₁ + χ₂) :=
+begin
+  apply weight_vector_multiplication L M M (to_module_hom R L M) χ₁ χ₂,
+  simp only [lie_module_hom.coe_to_linear_map, function.comp_app, linear_map.coe_comp,
+    tensor_product.map_incl, linear_map.mem_range],
+  use [⟨x, hx⟩ ⊗ₜ ⟨m, hm⟩],
+  simp only [submodule.subtype_apply, to_module_hom_apply, tensor_product.map_tmul],
+  refl,
+end
+
+variables (M)
+
+/-- If a Lie algebra is nilpotent, then pre-weight spaces are Lie submodules. -/
+def weight_space [lie_algebra.is_nilpotent R L] (χ : L → R) : lie_submodule R L M :=
+{ lie_mem := λ x m hm,
+  begin
+    rw ← zero_add χ,
+    refine lie_mem_pre_weight_space_of_mem_pre_weight_space _ hm,
+    suffices : pre_weight_space L (0 : L → R) = ⊤, { simp only [this, submodule.mem_top], },
+    exact lie_algebra.infi_max_gen_zero_eigenspace_eq_top_of_nilpotent R L,
+  end,
+  .. pre_weight_space M χ }
+
+@[simp] lemma mem_weight_space [lie_algebra.is_nilpotent R L] (χ : L → R) (m : M) :
```

Comment:
Are you sure you want the simp-lemma to go in this direction? Because it throws info away: the RHS is not about `lie_submodule`s.

Reply:
Great point. Definitely don't want this to be `simp`.